### PR TITLE
Semi Automatic Segmentation

### LINKED ISSUE #290: Semi automatic semantic segmentation

### PR DESCRIPTION
**Pull request by bsekachev**
Deep extreme cut works in backend.
More information about it: https://arxiv.org/pdf/1711.09081.pdf

Resolved #290

### CODE REVIEW

**Comment by nmanovic** [comment]
I cannot merge the PR anymore. Could you please merge develop branch into your branch?

File: cvat/apps/dextr_segmentation/dextr.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        if not self._plugin:
+            self._plugin = IEPlugin(device="CPU", plugin_dirs=[_IE_PLUGINS_PATH])
+            self._plugin.add_cpu_extension(os.path.join(_IE_PLUGINS_PATH, _IE_CPU_EXTENSION))
+            self._network = IENetwork.from_ir(model=self._xml, weights=self._bin)
+            self._input_blob = next(iter(self._network.inputs))
+            self._output_blob = next(iter(self._network.outputs))
+            self._exec_network = self._plugin.load(network=self._network)
+
+        image = PIL.Image.open(im_path)
+        numpy_image = np.array(PIL.Image.open(im_path))
```
Comment (savan77) [inline cvat/apps/dextr_segmentation/dextr.py]: Why are we reading the image again? We can copy `image` or use it directly if it isn't being used anywhere.

Comment (bsekachev) [reply cvat/apps/dextr_segmentation/dextr.py]: Yes, I agree.

**Comment by bsekachev** [comment]
@nmanovic 
Done.

File: cvat/apps/dextr_segmentation/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+This application uses the pre-trained DEXTR model which has been converted to Inference Engine format.
+
+We are grateful to K.K. Maninis, S. Caelles, J. Pont-Tuset, and L. Van Gool who permitted using their models in our tool
+
+## Installation
+
+```bash
+# Build image with Deep Extreme Cut application (OpenVINO component is also needed)
+docker-compose -f docker-compose.yml -f components/openvino/docker-compose.openvino.yml -f cvat/apps/dextr_segmentation/docker-compose.dextr.yml build
+```
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/README.md]: up -d?

File: cvat/apps/dextr_segmentation/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+# Semi-Automatic Segmentation with [Deep Extreme Cut](http://www.vision.ee.ethz.ch/~cvlsegmentation/dextr/)
+
+## About the application
+
+The application allows to use deep learning model for semi-automatic semantic and instance segmentation.
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/README.md]: deep learning models

File: cvat/apps/dextr_segmentation/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+You can get a segmentation polygon from four (or more) extreme points of object.
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/README.md]: of an object.

File: cvat/apps/dextr_segmentation/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+# Build image with Deep Extreme Cut application (OpenVINO component is also needed)
+docker-compose -f docker-compose.yml -f components/openvino/docker-compose.openvino.yml -f cvat/apps/dextr_segmentation/docker-compose.dextr.yml build
+```
+
+## Running
+
+1.  Open a job
+2.  Select "Auto Segmentation" in the list of shapes
+3.  Run the draw mode as usually (by press the "Create Shape" button or by "N" shortcut)
+4.  Click four-six (or more if need) extreme points of an object
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/README.md]: if it is needed

File: cvat/apps/dextr_segmentation/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+5.  Close the draw mode (the same way like start it)
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/README.md]: Close the draw mode as usually (by ...)

File: cvat/apps/dextr_segmentation/README.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+6.  Wait a moment and you will receive class agnostic annotation polygon
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/README.md]: will get a class ...

File: cvat/apps/dextr_segmentation/README.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+7.  You can close an annotation request if it is so long (in case if it is queued to rq worker and all workers are busy)
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/README.md]: if it is too long

File: cvat/apps/dextr_segmentation/dextr.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        if not _DEXTR_MODEL_DIR:
+            raise Exception("DEXTR_MODEL_DIR is not defined")
+        self._xml = os.path.join(_DEXTR_MODEL_DIR, 'dextr.xml')
+        self._bin = os.path.join(_DEXTR_MODEL_DIR, 'dextr.bin')
+
+    def handle(self, im_path, points):
+        # Lazy initialization
+        if not self._plugin:
+            self._plugin = IEPlugin(device="CPU", plugin_dirs=[_IE_PLUGINS_PATH])
+            self._plugin.add_cpu_extension(os.path.join(_IE_PLUGINS_PATH, _IE_CPU_EXTENSION))
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/dextr.py]: Why not the code below? Can be share inference between auto annotation and the app somehow?

```
        plugin = IEPlugin(device="CPU", plugin_dirs=[IE_PLUGINS_PATH])
        if (self._check_instruction("avx2")):
            plugin.add_cpu_extension(os.path.join(IE_PLUGINS_PATH, "libcpu_extension_avx2.so"))
        elif (self._check_instruction("sse4")):
            plugin.add_cpu_extension(os.path.join(IE_PLUGINS_PATH, "libcpu_extension_sse4.so"))
        else:
            raise Exception("Inference engine requires a support of avx2 or sse4.")
```

File: cvat/apps/dextr_segmentation/dextr.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        np.set_printoptions(threshold=np.nan)
+        pred = self._exec_network.infer(inputs={self._input_blob: input_dextr[np.newaxis, ...]})[self._output_blob][0, 0, :, :]
+        pred = cv2.resize(pred, tuple(reversed(numpy_cropped.shape[:2])), interpolation = cv2.INTER_CUBIC)
+        result = np.zeros(numpy_image.shape[:2])
+        result[bounding_box[1]:bounding_box[1] + pred.shape[0], bounding_box[0]:bounding_box[0] + pred.shape[1]] = pred > _DEXTR_TRESHOLD
+
+        # Convert a mask to a polygon
+        result = np.array(result, dtype=np.uint8)
+        cv2.normalize(result,result,0,255,cv2.NORM_MINMAX)
+        contours = None
```
Comment (nmanovic) [inline cvat/apps/dextr_segmentation/dextr.py]: Do we support two versions of OpenCV?

File: cvat/apps/dextr_segmentation/dextr.py (hunk 9 above)
Comment (bsekachev) [reply cvat/apps/dextr_segmentation/dextr.py]: Yes, I agree
I have done it.

File: cvat/apps/dextr_segmentation/dextr.py (hunk 10 above)
Comment (bsekachev) [reply cvat/apps/dextr_segmentation/dextr.py]: Yes
It works with both OpenCV versions

**Pull request merged**